### ISSUE TITLE
XLSR-53 crashes with NaNs after some epochs of pretraining

### ISSUE THREAD

**Issue by gowtham1997** [pytorch/fairseq#3750]
### Before asking:   
1. search the issues.   
2. search the docs.    

#### What is your question?

We tried pretraining the XLSR-53 checkpoint on some newer languages dataset of around 10K hours. The training loss seems to explode after few epochs and lowering the learning rate, increasing the batch size, etc doesn't seem to help (it delays the exploding gradients to 3-4 more epochs but the training always seems to crash).

```
[2021-07-22 17:10:50,251][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.5
[2021-07-22 17:15:02,126][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.25
[2021-07-22 17:28:47,978][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.25
[2021-07-22 17:45:07,100][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.25
[2021-07-22 18:09:10,411][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.5
[2021-07-22 18:19:29,514][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.5
[2021-07-22 18:23:28,679][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.25
[2021-07-22 18:37:12,573][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.25
[2021-07-22 18:59:19,864][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.5
[2021-07-22 19:07:13,025][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.25
[2021-07-22 19:07:22,087][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.125
[2021-07-22 19:07:31,087][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.0625
[2021-07-22 19:07:44,846][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.03125
[2021-07-22 19:07:58,494][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.015625
[2021-07-22 19:08:03,117][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.0078125
[2021-07-22 19:08:07,374][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.00390625
[2021-07-22 19:08:16,389][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.001953125
[2021-07-22 19:08:25,450][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.0009765625
[2021-07-22 19:08:30,198][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.00048828125
[2021-07-22 19:08:34,115][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.000244140625
[2021-07-22 19:08:38,548][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 0.0001220703125
```

Additionally, we implement temperature sampling and keep `alpha` to 0.7 

#### Code

   

We use the same config as [wav2vec2 large model](https://github.com/pytorch/fairseq/blob/master/examples/wav2vec/config/pretraining/wav2vec2_large_librivox.yaml)

^ We increase the `max_tokens` from 1200000 to 1800000.

#### What have you tried?

1) training with LR 5e-4 and warmup_steps of 32K:
![image](https://user-images.githubusercontent.com/9641196/127442015-38b3f742-8309-4af4-b220-657ce541b15d.png)

^ the loss decreases as expected for 2 epochs and at the start of epoch 3, the training crashes due to exploding gradients.

2) training with LR 5e-4 and warmup_steps of 32K and clipping l2 norm of gradients to 1.0
![image](https://user-images.githubusercontent.com/9641196/127442304-1f8ae7f8-f494-4d80-94a5-f4f8bafe4726.png)

^ similar behavior as before.

3) training with LR 5e-5 and no warmup_steps and clipping l2 norm of gradients to 1.0
![image](https://user-images.githubusercontent.com/9641196/127442460-d8fd2841-968f-4d05-9278-93dabf91c45f.png)

^ This runs for 10 epochs before crashing due to the higher L2 norm of the gradients.
![image](https://user-images.githubusercontent.com/9641196/127442557-691de464-8305-4c4f-8ad6-b2c774a916aa.png)

4) training with LR 5e-5 and no warmup_steps and clipping l2 norm of gradients to 0.5

^ similar behavior as 3) but runs for one more epoch.

The best checkpoints obtained from experiment 3) or 4) also crash with exploding gradients during fine-tuning 

Any ideas on how to make the training stable? @alexeib @aconneau @myleott @michaelauli 

#### What's your environment?

 - fairseq Version (e.g., 1.0 or master): master
 - PyTorch Version (e.g., 1.0) 1.7.1
 - OS (e.g., Linux): Linnux
 - How you installed fairseq (`pip`, source): source
 - Build command you used (if compiling from source): `pip install -e .`
 - Python version: 3.7
 - CUDA/cuDNN version: 11.0
 - GPU models and configuration: 8 A100s 
 - Any other relevant information:

**Comment by codecaution** [comment]
I meet the same problem. 
Do you have any ideas to fix it now? It seems come from the model side, which generates large gradient tensor value and then overflow.

**Comment by gowtham1997** [comment]
@codecaution Haven't solved it yet. We tried switching to fp32 training but that runs with smaller batch sizes and was very slow.

**Comment by codecaution** [comment]
@gowtham1997 How about BF16 in A100?

**Comment by gowtham1997** [comment]
@codecaution Thanks for this suggestion. We will try this out

But if the gradients are exploding in fp16, why would this behavior change with bf16?

**Comment by alexeib** [comment]
add a non affine layer norm on the output of post_concat_proj, project_q and change "layer_norm" to be non affine and it should train fine but at a cost of some accuracy (which i guess you'll make up for by not crashing)

**Comment by codecaution** [comment]
@gowtham1997 The loss scale is even 0.0001220703125, which maybe mean your gradient/loss is very large? The max value of FP16 is 65504, while BF16 is 3e38.

**Comment by firdota** [comment]
hello, do you have any progress? @gowtham1997 @codecaution

**Comment by gowtham1997** [comment]
@firdota I haven't tried layer norm or bf16 yet. But planning to do it this week and will update soon

**Comment by codecaution** [comment]
@gowtham1997 @firdota From my side, both BF16 and FP32 can work in my model.

**Comment by gowtham1997** [comment]
> add a non affine layer norm on the output of post_concat_proj, project_q and change "layer_norm" used on "features" to be non-affine and it should train fine but at a cost of some accuracy (which i guess you'll make up for by not crashing)

@alexeib Added non trainable layernorm to project_q, post_extract_proj (couldn't find post_concat_proj, I think you meant post_extract_proj). Also changed the [layernorm](https://github.com/pytorch/fairseq/blob/9825786fbe8a32053f21bec988d953d175f7262a/fairseq/models/wav2vec/wav2vec2.py#L538) on features to be non-affline.

Should we also add another non affine layer norm for output of [project_inp](https://github.com/pytorch/fairseq/blob/9825786fbe8a32053f21bec988d953d175f7262a/fairseq/models/wav2vec/wav2vec2.py#L580) ?

**Comment by alexeib** [comment]
i was able to stabilize most models with just the changes i mentioned, but you are free to experiment and please share the results!

**Comment by firdota** [comment]
> @firdota I haven't tried layer norm or bf16 yet. But planning to do it this week and will update soon

look forward to your good news

**Comment by firdota** [comment]
> @gowtham1997 @firdota From my side, both BF16 and FP32 can work in my model.

great!

**Comment by lixx0105** [comment]
> > add a non affine layer norm on the output of post_concat_proj, project_q and change "layer_norm" used on "features" to be non-affine and it should train fine but at a cost of some accuracy (which i guess you'll make up for by not crashing)
> 
> @alexeib Added non trainable layernorm to project_q, post_extract_proj (couldn't find post_concat_proj, I think you meant post_extract_proj). Also changed the [layernorm](https://github.com/pytorch/fairseq/blob/9825786fbe8a32053f21bec988d953d175f7262a/fairseq/models/wav2vec/wav2vec2.py#L538) on features to be non-affline.
> 
> Should we also add another non affine layer norm for output of [project_inp](https://github.com/pytorch/fairseq/blob/9825786fbe8a32053f21bec988d953d175f7262a/fairseq/models/wav2vec/wav2vec2.py#L580) ? (I'm guessing this is only useful when inputs need to be quantized and currently its False in case of XLSR)

hi @gowtham1997 @codecaution  I'm facing simliar problem. I pretrain the model with my own 10k hours audio datasets and initialize the parameters with facebook released model.  May i know your solutions? Does bf16/fp32 or the non-trainable layernorm work in your test? thank you.

**Comment by gowtham1997** [comment]
We tried training with nontrainable layer norms but the performance seems to drop a lot.

![image](https://user-images.githubusercontent.com/9641196/129710424-ac5ebfbf-9861-46bf-b572-e783de008583.png)

^ 
grey is loss plot Wav2vec-base model trained on the same data
light orange is the fairseq xlsr checkpoint (LR 5e-5) where we faced NaN problem (this starts with a much smaller loss and then runs into NANs)

Yellow is the loss plot of XLSR with nontrainable layer norms and LR 5e-3. There are no NAN problems with this model but the performance was comparatively worse and also behind the wav2vec base model.

Light blue is loss plot for XLSR with nontrainable layer norms and LR 5e-5. It seems to be better than yellow but still runs into NANs after some epochs of training.

**Comment by lixx0105** [comment]
@gowtham1997 thanks for your quick response. l'm not sure whether the loss during pretraining is directly related to the performace of downstream asr. My case is actually using wav2vec-base model as initialization. the gradients are overflow after several epoch.
